axiom/run.go

```diff
@@ -0,0 +1,258 @@
+			if failfast {
+				return err
+			}
+			log.Printf("benchmark error: %v", err)
+		}
+		id++
+	}
+
+	return nil
+}
+
+func gitSha() string {
+	sha, err := exec.Command("git", "rev-parse", "--short", "HEAD").Output()
+	if err != nil {
+		return ""
+	}
+	return strings.TrimSpace(string(sha))
+}
+
+func benchmark(ctx context.Context, cli *axiomClient, id int, query string, iters int, enc *json.Encoder) error {
+	for i := 1; i <= iters; i++ {
+		result, err := cli.Query(ctx, id, query)
+		if err != nil {
+			return fmt.Errorf("failed query #%d, iter %d: %w", id, i, err)
+		}
+
+		if err = enc.Encode(result); err != nil {
+			return fmt.Errorf("failed to encode result of query #%d, iter %d: %w", id, i, err)
+		}
+
+		// We want to encode out results with errors, but still return early if there was
+		// one.
+		if result.Error != "" {
+			return fmt.Errorf("failed query #%d, iter %d: %s", id, i, result.Error)
+		}
+	}
+
+	return nil
+}
+
+type axiomClient struct {
+	cli      *http.Client
+	apiURL   *url.URL
+	traceURL *url.URL
+	version  string
+	token    string
+	org      string
+}
+
+func newAxiomClient(cli *http.Client, version, apiURL, org, token, traceURL string) (*axiomClient, error) {
+	parsedTraceURL, err := url.Parse(traceURL)
+	if err != nil && traceURL != "" {
+		return nil, fmt.Errorf("error parsing trace url: %w", err)
+	}
+
+	parsedAPIURL, err := url.Parse(apiURL)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing url: %w", err)
+	}
+
+	return &axiomClient{
+		cli:      cli,
+		apiURL:   parsedAPIURL,
+		traceURL: parsedTraceURL,
+		version:  version,
+		token:    token,
+		org:      org,
+	}, nil
+}
+
+type QueryResult struct {
+	// Query is the APL query submitted
+	Query string `json:"query"`
+	// ID is the clickbench query number [1-43]
+	ID int `json:"id"`
+	// URL of the query request. May include query arguments like nocache=true
+	URL string `json:"url"`
+	// Time is the time the query was submitted
+	Time time.Time `json:"_time"`
+	// LatencyNanos is the total latency of the query in nanoseconds, including network round-trips.
+	LatencyNanos time.Duration `json:"latency_nanos"`
+	// LatencySeconds is the total latency of the query in seconds, including network round-trips.
+	LatencySeconds float64 `json:"latency_seconds"`
+	// ServerVersions is a dictionary of service name to git sha that was under test
+	ServerVersions map[string]string `json:"server_versions"`
+	// Version is the git sha of the benchmarking client code
+	Version string `json:"version"`
+	// TraceID is the trace ID of the query request
+	TraceID string `json:"trace_id"`
+	// TraceURL is the URL to the trace in Axiom
+	TraceURL string `json:"trace_url"`
+	// Result is the query result
+	Result *query.Result `json:"result"`
+	// Error is the error if the query failed
+	Error string `json:"error"`
+}
+
+func (c *axiomClient) Query(ctx context.Context, id int, aplQuery string) (*QueryResult, error) {
```

Comment:
it's not clear to me why we need a custom Axiom client, and can't use the one from the axiom-go lib that we import?

Reply:
Nope, because it doesn't expose the trace-id from the headers, and designing a solution to that generated some [discussion internally](https://watchlyhq.slack.com/archives/C03C9UL1UE8/p1694180174064299) which might pan out later, but for now we shouldn't block.